components/script/script_thread.rs

```diff
@@ -2259,35 +2186,27 @@ impl Drop for ScriptThread {
     }
 }
 
-/// Shuts down layout for the given browsing context tree.
-fn shut_down_layout(context_tree: &BrowsingContext) {
-    let mut channels = vec!();
-
-    for context in context_tree.iter() {
-        // Tell the layout thread to begin shutting down, and wait until it
-        // processed this message.
-        let (response_chan, response_port) = channel();
-        let window = context.active_window();
-        let chan = window.layout_chan().clone();
-        if chan.send(message::Msg::PrepareToExit(response_chan)).is_ok() {
-            channels.push(chan);
-            let _ = response_port.recv();
-        }
+/// Shuts down layout for the given window.
+fn shut_down_layout(document: &Document) {
+    // Tell the layout thread to begin shutting down, and wait until it
+    // processed this message.
+    let (response_chan, response_port) = channel();
+    let chan = document.window().layout_chan().clone();
+    if chan.send(message::Msg::PrepareToExit(response_chan)).is_ok() {
+        let _ = response_port.recv();
     }
 
     // Drop our references to the JSContext and DOM objects.
-    for context in context_tree.iter() {
-        let window = context.active_window();
-        window.clear_js_runtime();
+    document.window().clear_js_runtime();
 
-        // Sever the connection between the global and the DOM tree
+    // If the window is for an active document, then
+    // sever the connection between the global and the DOM tree
```

Comment:
The function is odd - `document.window()` can return a Window for a document that has no browsing context, but that doesn't mean that we want to shut down layout for that window. Perhaps this function should accept a `&Window` argument instead, and obtain the context from that.